src/SecTester.Repeater/Repeater.cs

```diff
@@ -1,37 +1,163 @@
+    {
+      Status = RunningStatus.Off;
+      throw;
+    }
+    finally
+    {
+      ReleaseLock();
+    }
+  }
+
+  private async Task SchedulePing()
+  {
+    await SendStatus(RepeaterStatus.Connected).ConfigureAwait(false);
+    _heartbeat.Elapsed += async (_, _) => await SendStatus(RepeaterStatus.Connected).ConfigureAwait(false);
+    _heartbeat.Start();
+  }
+
+  private async Task SendStatus(RepeaterStatus status)
+  {
+    var @event = new RepeaterStatusEvent(RepeaterId, status);
+    await _eventBus.Publish(@event).ConfigureAwait(false);
+  }
+
+  private async Task Register()
+  {
+    var command = new RegisterRepeaterCommand(_version.ToString(), RepeaterId);
+    var res = await _eventBus.Execute(command).ConfigureAwait(false);
+
+    if (res == null)
+    {
+      throw new SecTesterException("Error registering repeater.");
+    }
+
+    EnsureRegistrationStatus(res);
+  }
+
+  public async Task Stop(CancellationToken cancellationToken = default)
+  {
+    try
+    {
+      await _semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+      if (Status != RunningStatus.Running)
+      {
+        return;
+      }
+
+      Status = RunningStatus.Off;
+
+      _heartbeat.Stop();
+      await SendStatus(RepeaterStatus.Disconnected).ConfigureAwait(false);
+      // TODO: dispose an event bus
+    }
+    finally
+    {
+      ReleaseLock();
+    }
+  }
+
+  private void EnsureRegistrationStatus(RegisterRepeaterResult result)
+  {
+    if (result.Error != RepeaterRegisteringError.None)
+    {
+      HandleRegisterError(result.Error);
+    }
+    else
+    {
+      if (new Version(result.Version!).CompareTo(_version) != 0)
+      {
+        // TODO: colorize an output in the same manner like sectester-js does
+        _logger.LogWarning(
+          "(!) IMPORTANT: A new Repeater version ({Version}) is available, please update SecTester.",
+          result.Version
+        );
+      }
+    }
+  }
+
+  private void HandleRegisterError(RepeaterRegisteringError error)
   {
-    throw new NotImplementedException();
+    throw error switch
+    {
+      RepeaterRegisteringError.NotActive => new SecTesterException("Access Refused: The current Repeater is not active."),
+      RepeaterRegisteringError.NotFound => new SecTesterException("Unauthorized access. Please check your credentials."),
+      RepeaterRegisteringError.Busy => new SecTesterException(
+        $"Access Refused: There is an already running Repeater with ID {RepeaterId}"),
+      RepeaterRegisteringError.RequiresToBeUpdated => new SecTesterException(
+        "(!) CRITICAL: The current running version is no longer supported, please update SecTester."),
+      _ => new ArgumentOutOfRangeException(nameof(error), error, "Something went wrong. Unknown error.")
+    };
   }
 
-  public Task Stop()
+  private void ReleaseLock()
   {
-    throw new NotImplementedException();
+    try
+    {
+      _semaphore.Release();
+    }
+    catch
+    {
+      // ignored
+    }
```

Comment:
This try/finally based locking does not provide clear expectation whether semaphore lock is obtained or not.
E.g. having already canceled token in the Start method lead to `Release` on an unlocked instance. Instead of this, I suggest switching to `Disposable` pattern locking.

```csharp
internal class SemaphoreLock : IDisposable
{
  private readonly SemaphoreSlim _semaphore;

  private SemaphoreLock(SemaphoreSlim semaphore)
  {
    _semaphore = semaphore ?? throw new ArgumentNullException(nameof(semaphore));
  }

  public static async Task<IDisposable> LockAsync(SemaphoreSlim semaphore, CancellationToken cancellationToken = default)
  {
    var releaser = new SemaphoreLock(semaphore);

    await semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);

    return releaser;
  }

  public static IDisposable Lock(SemaphoreSlim semaphore)
  {
    var releaser = new SemaphoreLock(semaphore);

    semaphore.Wait();

    return releaser;
  }

  public void Dispose()
  {
    _semaphore.Release();
    GC.SuppressFinalize(this);
  }
}
```

Reply:
@ostridm well, especially for that, to handle both cases when a lock has been already released or even when it is still not acquired, we wrap `Release` to try/catch. `Disposable` pattern does not solve this issue, just provides a more elegant way of reducing nesting.